common/crypto/dkg/src/bte/proof_chunking.rs

```diff
@@ -0,0 +1,1084 @@
+                // scalar in range of [0, Z - 1 + S]
+                let shifted = rng.gen_range(0..=combined_upper_range);
+                // [-S, Z - 1] as required
+                // TODO: CHECK FOR OFF BY ONE ERRORS
+                let blinding_factor = Scalar::from(shifted) - ss_scalar;
+
+                // y0 ^ beta_l • g1 ^ sigma_l
+                let cc = y0 * betas[i] + g1 * blinding_factor;
+
+                // blinding_factors.push(blinding_factor);
+                shifted_blinding_factors.push(shifted);
+                cs.push(cc);
+            }
+
+            let first_challenge = Self::compute_first_challenge(&instance, &y0, &bs, &cs, n, m);
+
+            // compute:
+            // z_{s,1} = sum(i <- 1..n; (sum j <- j..m; e_{i,j,1} • s_{i,j} + sigma_1))
+            // ...
+            // z_{s,l} = sum(i <- 1..n; (sum j <- j..m; e_{i,j,l} • s_{i,j} + sigma_l))
+            // such that 0 <= z_{s,l} < Z
+            let mut responses_chunks = Vec::with_capacity(PARALLEL_RUNS);
+
+            // I think this part is more readable with a range loop
+            #[allow(clippy::needless_range_loop)]
+            for l in 0..PARALLEL_RUNS {
+                let mut sum = 0;
+
+                for (i, witness_i) in witnesses_s.iter().enumerate() {
+                    for (j, witness_ij) in witness_i.to_chunks().chunks.iter().enumerate() {
+                        debug_assert!(std::mem::size_of::<Chunk>() <= std::mem::size_of::<u64>());
+                        sum += first_challenge[i][j][l] * (*witness_ij as u64)
+                    }
+                }
+
+                if sum + shifted_blinding_factors[l] < ss {
+                    continue 'retry_loop;
+                }
+                // shifted_blinding_factors[l] - ss restores it to "proper" [-S, Z - 1] range
+                let response = sum + shifted_blinding_factors[l] - ss;
+                if response < zz {
+                    responses_chunks.push(response)
+                } else {
+                    continue 'retry_loop;
+                }
+            }
+
+            break (first_challenge, responses_chunks, cs);
+        };
+
+        let mut deltas = Vec::with_capacity(n + 1);
+        let mut ds = Vec::with_capacity(n + 1);
+
+        for _ in 0..=n {
+            let delta = Scalar::random(&mut rng);
+            let dd = g1 * delta;
+
+            deltas.push(delta);
+            ds.push(dd);
+        }
+
+        //  Y = y_0 ^ delta_0 • y_1 ^ delta_1 • ... • y_n ^ delta_n
+        let mut yy = y0 * deltas[0];
+        for (i, delta_i) in deltas.iter().enumerate().skip(1) {
+            yy += instance.public_keys[i - 1].0 * delta_i;
+        }
+
+        let second_challenge =
+            Self::compute_second_challenge(&first_challenge, &responses_chunks, &ds, &yy);
+
+        // compute responses
+
+        // TODO: is this correct capacity?
+        let mut responses_r = Vec::with_capacity(PARALLEL_RUNS);
```

Comment:
Shouldn't there be n of those responses?

Reply:
Yup, you're totally right -> adjusted.